nightfall-deployer/migrations/4_upgrade_upgradeable.js

```diff
@@ -0,0 +1,32 @@
+const { deployProxy } = require('@openzeppelin/truffle-upgrades');
+
+const Verifier = artifacts.require('Verifier.sol');
+const Shield = artifacts.require('Shield.sol');
+const MerkleTree_Stateless = artifacts.require('MerkleTree_Stateless.sol');
+const MiMC = artifacts.require('MiMC.sol');
+const Structures = artifacts.require('Structures.sol');
+const Config = artifacts.require('Config.sol');
+const Utils = artifacts.require('Utils.sol');
+const ChallengesUtil = artifacts.require('ChallengesUtil.sol');
+const Proposers = artifacts.require('Proposers.sol');
+const Challenges = artifacts.require('Challenges.sol');
+const State = artifacts.require('State.sol');
+
+module.exports = async function(deployer) {
+  await deployer.deploy(Verifier);
+  await deployer.link(Verifier, [Challenges,ChallengesUtil]);
+  await deployer.deploy(MiMC);
+  await deployer.link(MiMC, MerkleTree_Stateless);
+  await deployer.deploy(MerkleTree_Stateless);
+  await deployer.link(MerkleTree_Stateless, [Challenges,ChallengesUtil]);
+    await deployer.deploy(Utils);
+  await deployer.link(Utils, [Shield,Challenges,ChallengesUtil]);
+  await deployer.deploy(ChallengesUtil);
+  await deployer.link(ChallengesUtil, Challenges);
+  await deployProxy(Structures, [], {deployer, unsafeAllowLinkedLibraries: true });
```

Comment:
This is the same file as `nightfall-deployer/migrations/2_deploy_upgradeable.js`.
I am assuming this is WIP that will be tied up in upgrade PR 2, which will use `upgradeProxy` instead of `deployProxy` here?

Reply:
All of these are libraries, i.e. they contain only pure functions and they're only called by other contracts. Thus there didn't seem to be any point in proxying them because we can just replace them if we need to, and link the new versions to the more stateful contracts.
I may change my mind but that's my current thinking.